src/Compilers/CSharp/Portable/Declarations/MergedNamespaceDeclaration.cs

```diff
@@ -139,43 +140,100 @@ private ImmutableArray<MergedNamespaceOrTypeDeclaration> MakeChildren()
-                }
-                else
+                if (types != null)
                 {
-                    var typeGroups = types.ToDictionary(t => t.Identity);
-                    types.Free();
-
-                    foreach (var typeGroup in typeGroups.Values)
+                    if (allTypesHaveSameIdentity)
                     {
-                        children.Add(new MergedTypeDeclaration(typeGroup));
+                        children.Add(new MergedTypeDeclaration(types.ToImmutableAndFree()));
+                    }
+                    else
+                    {
+                        // PERF: Use object as the value in this dictionary to efficiently represent single item collections.
+                        // If only a single object has been seen with a given identity, the value will be a SingleTypeDeclaration,
+                        // otherwise, the value will be an ArrayBuilder<SingleTypeDeclaration>. This code differs from
+                        // addNamespacesToChildren intentionally as the vast majority of identities are represented by only a
+                        // single item in the types collection.
+                        var typeGroups = PooledDictionary<SingleTypeDeclaration.TypeDeclarationIdentity, object>.GetInstance();
+
+                        foreach (var t in types)
+                        {
+                            var id = t.Identity;
+
+                            if (typeGroups.TryGetValue(id, out var existingValue))
+                            {
+                                if (existingValue is not ArrayBuilder<SingleTypeDeclaration> builder)
+                                {
+                                    builder = ArrayBuilder<SingleTypeDeclaration>.GetInstance();
+                                    builder.Add((SingleTypeDeclaration)existingValue);
+                                    typeGroups[id] = builder;
+                                }
+
+                                builder.Add(t);
+                            }
+                            else
+                            {
+                                typeGroups.Add(id, t);
+                            }
+                        }
+
+                        foreach (var typeGroup in typeGroups.Values)
+                        {
+                            if (typeGroup is SingleTypeDeclaration t)
+                            {
+                                children.Add(new MergedTypeDeclaration([t]));
+                            }
+                            else
+                            {
+                                var builder = (ArrayBuilder<SingleTypeDeclaration>)typeGroup;
+                                children.Add(new MergedTypeDeclaration(builder.ToImmutableAndFree()));
+                            }
+                        }
+
+                        types.Free();
+                        typeGroups.Free();
```

Comment:
doc that we don't need to free the builder values as that was haneled in the loop above?